Mathlib/Tactic/DeriveInfinite.lean

```diff
@@ -0,0 +1,350 @@
+Authors: Timothy Gu
+-/
+import Lean
+import Mathlib.Data.Fintype.Card
+
+/-!
+# `Infinite` derive handler
+
+This module defines a `Infinite` derive handler for inductive types.
+The handler works if the inductive type `α` satisfies one of the following:
+
+1. There is a constructor whose fields are all `Nonempty` and has at least one `Infinite` field.
+   (Note that `Infinite` implies `Nonempty`.)
+2. `α` is `Inhabited`, AND there is a constructor whose fields are all `Nonempty` and has at least
+   one `α`-valued field.
+   (Note that `Inhabited` implies `Nonempty`, but not vice versa.)
+
+Here are two examples of what this handler can handle:
+```
+structure S where
+  a : Nat
+  b : Unit
+  deriving Infinite
+
+inductive Nat' where
+  | zero : Nat'
+  | succ (n : Nat') : Nat'
+  deriving Inhabited, Infinite
+```
+
+Additionally, this module provides the `derive_infinite%` term elaborator, which allows adding some
+more hypotheses to the derived `Infinite` instance.
+```
+inductive List' (α : Type u) where
+  | nil : List' α
+  | cons (head : α) (tail : List' α) : List' α
+  deriving Inhabited
+
+instance [Nonempty α] : Infinite (List' α) := derive_infinite% _
+```
+
+The deriving handler does not currently support mutually inductive types. It also doesn't support
+indexed inductive types in general, but `derive_infinite%` works for specific indexed types:
+```
+inductive Vector' (α : Type u) : Nat → Type u where
+  | nil : Vector' α 0
+  | cons (head : α) (tail : Vector' α n) : Vector' α (n + 1)
+instance [Inhabited α] : ∀ n, Inhabited (Vector' α n) := ⋯
+instance [Inhabited α] [Infinite α] : Infinite (Vector' α (n + 1)) := derive_infinite% _
+```
+
+## Implementation Notes
+
+To get debugging information (including generated proof), run
+`set_option trace.Elab.Deriving.infinite true`.
+
+The generated proof uses one of the following two strategies, depending on which of the two
+scenarios we are currently in.
+1. If there is a constructor with an `Infinite` field, we try to construct a proof for
+   `Infinite α` by forming an injection from that `Infinite` type to `α`, and then using the
+   lemma `Infinite.of_injective`.
+2. If there is a constructor with an `α`-valued field, we try to construct a proof for
+   `Infinite α` by forming an injection from `α` to the subset of `α` that is constructed using
+   this constructor. Hopefully, the `default` value of `α` uses another constructor, which would
+   show that this subset is proper. The lemma `Infinite.of_injective_to_set` then completes the
+   proof.
+
+The deriving handler is implemented in terms of `derive_infinite%` in the following way (modulo
+inductive type parameters):
+```
+instance : Infinite α := derive_infinite% _
+```
+-/
+
+namespace Mathlib.Deriving.Infinite
+open Lean Elab
+open Lean.Elab.Command
+open Lean.Parser.Term
+open Lean.Meta
+
+/--
+Given a constructor `ctor` with `m` inductive parameters, `n` fields, and `fieldIdx = i`, returns
```

Comment:
```suggestion
Given a constructor `ctor` with `m` inductive parameters, `n` fields, and `fieldIdx = i`, returns
```
I think we usually say `argument` rather than `field`. (The phrase below "infinite field" is particularly confusing to a mathematician's ears :-)